Corona-Warn-App/src/main/java/de/rki/coronawarnapp/ui/view/CountryList.kt

```diff
@@ -2,47 +2,98 @@ package de.rki.coronawarnapp.ui.view
         }
 
     init {
-        orientation = VERTICAL
+        orientation = HORIZONTAL
     }
 
     /**
      * Cleans the view and rebuilds the list of countries. Presets already selected countries
      */
-    private fun buildList() {
-        this.removeAllViews()
+    private fun buildAllLists() {
         list
             ?.map { country ->
                 context.getString(country.labelRes) to country.iconRes
             }
             ?.sortedWith { a, b ->
                 Collator.getInstance().compare(a.first, b.first)
             }
-            ?.forEachIndexed { index, (label, iconRes) ->
-                inflate(context, R.layout.view_country_list_entry, this)
-                val child = this.getChildAt(index)
-                child.apply {
-                    findViewById<ImageView>(R.id.country_list_entry_image).setImageResource(iconRes)
-                    findViewById<TextView>(R.id.country_list_entry_label).text = label
+
+        view = inflate(context, R.layout.view_country_list_entry_flag_container, this)
+        val flagNumberOfColumns = 8
+        var adapterCountryFlags = CountryAdapterFlags(context, _list)
+        context.withStyledAttributes(attributeSet, R.styleable.SimpleStepEntry) {
+
+            flagGrid.layoutManager = GridLayoutManager(context, flagNumberOfColumns)
+            flagGrid.adapter = adapterCountryFlags
+
+            var countryNames = ""
+            val iterator: Iterator<Country> = list!!.iterator()
+            while (iterator.hasNext()) {
+                countryNames += resources.getString(iterator.next().labelRes)
+                if (iterator.hasNext()) {
+                    countryNames += ", "
                 }
             }
+            country_list_entry_label.text = countryNames
+        }
+    }
+}
+
+/**
+ * Adapter for the country flags grid
+ */
+class CountryAdapterFlags internal constructor(
+    context: Context?,
+    countryList: List<Country>?
+) :
+    RecyclerView.Adapter<CountryAdapterFlags.ViewHolder>() {
+    private val mInflater: LayoutInflater = LayoutInflater.from(context)
+    private val countryList: List<Country>? = countryList
+
+    // inflates the cell layout from xml when needed
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
+        val view: View = mInflater.inflate(R.layout.view_country_list_entry_flag_item, parent, false)
+        return ViewHolder(view)
+    }
+
+    // binds the data to the TextView in each cell
+    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
+        val country = this.countryList!![position]
+        holder.flagImage.setImageResource(country.iconRes)
+    }
+
+    // total number of cells
+    override fun getItemCount(): Int {
+        return countryList!!.size
+    }
+
+    // stores and recycles views as they are scrolled off screen
+    inner class ViewHolder internal constructor(itemView: View) : RecyclerView.ViewHolder(itemView) {
```

Comment:
Check out `KeyFileDownloadAdapter.kt` as example implementation, it does not have to be as elaborate (but would be nice), the gist is to delegate the view specific details (inflation & binding) to the `ViewHolder` and to take as few dependencies via constructor (i.e. no `Context'), if we can get it from the RecyclerView (i.e. getting a context from the `parent: ViewGroup`).

https://github.com/corona-warn-app/cwa-app-android/blob/ec910ab5c8c156d069c2b7c384b1d1c40dfd21ce/Corona-Warn-App/src/deviceForTesters/java/de/rki/coronawarnapp/test/keydownload/ui/KeyFileDownloadAdapter.kt#L15

Pass the parent to the ViewHolder, let the ViewHolder inflate it's own layout (modularization), and only pass the data class (here `Country`) to a bind method in the ViewHolder.

Reply:
Please check the new implmentation out when you can